web/interceptor/tokens.go

```diff
@@ -0,0 +1,75 @@
+package interceptor
+
+import (
+	"context"
+	"log"
+	"strings"
+
+	"github.com/quickfeed/quickfeed/web/auth"
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+)
+
+type (
+	// The userIDs interface must be implemented by request types that may need to update the tokens.
+	userIDs interface{ UserIDs() []uint64 }
+	// Marker interface to detect the GroupRequest type needed for DeleteGroup.
+	isGroup interface{ GetGroupID() uint64 }
+)
+
+var defaultTokenUpdater = func(_ context.Context, tm *auth.TokenManager, msg userIDs) error {
+	for _, userID := range msg.UserIDs() {
+		if err := tm.Add(userID); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// tokenUpdateMethods is a map of methods that require updating the list of users who need a new JWT.
+var tokenUpdateMethods = map[string]func(context.Context, *auth.TokenManager, userIDs) error{
+	"UpdateUser":        defaultTokenUpdater, // User has been promoted to admin or demoted.
+	"UpdateGroup":       defaultTokenUpdater, // Users added to a group or removed from a group.
+	"UpdateEnrollments": defaultTokenUpdater, // User enrolled into a new course or promoted to TA.
+
+	"CreateCourse": // The signed in user gets the teacher role in the new course.
+	func(ctx context.Context, tm *auth.TokenManager, _ userIDs) error {
+		claims, err := tm.GetClaims(ctx)
+		if err != nil {
+			return err
+		}
+		return tm.Add(claims.UserID)
+	},
+
+	"DeleteGroup": // Group members removed from the group.
+	func(ctx context.Context, tm *auth.TokenManager, msg userIDs) error {
+		if grp, ok := msg.(isGroup); ok {
+			group, err := tm.Database().GetGroup(grp.GetGroupID())
+			if err != nil {
+				return err
+			}
+			return defaultTokenUpdater(ctx, tm, group)
+		}
+		return nil // TODO(meling) This should probably be an error??
+	},
+}
+
+// TokenRefresher updates list of users who need a new JWT next time they send a request to the server.
+// This method only logs errors to avoid overwriting the gRPC error messages returned by the server.
+func TokenRefresher(logger *zap.SugaredLogger, tm *auth.TokenManager) grpc.UnaryServerInterceptor {
+	return func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
+		method := info.FullMethod[strings.LastIndex(info.FullMethod, "/")+1:]
+		if tokenUpdateFn, ok := tokenUpdateMethods[method]; ok {
+			if msg, ok := req.(userIDs); ok {
+				log.Println("REFRESHER: ", method) // tmp
+
+				if err := tokenUpdateFn(ctx, tm, msg); err != nil {
+					logger.Error(err)
```

Comment:
Should this return an error or fall through to 73 and `handler`? Same question for Line 70.

Reply:
I'm a bit conflicted about returning errors from this interceptor, but probably yes, it is better to return an error in both cases.